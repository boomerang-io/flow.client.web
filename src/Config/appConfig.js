// Look for the data injected into the HTML file from the Express app
// See server/app.js for implementation
export const APP_ROOT =
  window._SERVER_DATA && window._SERVER_DATA.APP_ROOT ? window._SERVER_DATA.APP_ROOT : "/BMRG_APP_ROOT_CONTEXT";

export const PRODUCT_STANDALONE =
  window._SERVER_DATA && window._SERVER_DATA.PRODUCT_STANDALONE
    ? window._SERVER_DATA.PRODUCT_STANDALONE === "true"
    : process.env.REACT_APP_PRODUCT_STANDALONE
    ? true
    : false;

export const EMBEDDED_MODE =
  window._SERVER_DATA && window._SERVER_DATA.EMBEDDED_MODE
    ? window._SERVER_DATA.EMBEDDED_MODE === "true"
    : process.env.REACT_APP_EMBEDDED_MODE
    ? true
    : true;

export const CORE_ENV_URL =
  window._SERVER_DATA && window._SERVER_DATA.CORE_ENV_URL ? window._SERVER_DATA.CORE_ENV_URL : "";

export const BASE_DOCUMENTATION_URL = PRODUCT_STANDALONE
  ? "https://www.useboomerang.io/docs/boomerang-flow"
  : `${CORE_ENV_URL}/docs/boomerang-flow`;

export const isDevEnv = process.env.NODE_ENV === "development";
export const isTestEnv = process.env.NODE_ENV === "test";

export const AppPath = {
  Root: "/",
  Error: "/error",
  Activity: "/activity",
  Editor: "/editor/:workflowId",
  EditorDesigner: `/editor/:workflowId/workflow`,
  EditorConfigure: `editor/:workflowId/configure`,
  EditorChangelog: `/editor/:workflowId/changelog`,
  EditorProperties: `/editor/:workflowId/parameters`,
  Execution: "/activity/:workflowId/execution/:executionId",
  Insights: "/insights",
  Properties: "/admin/parameters",
  Quotas: "/admin/quotas",
  QuotasEdit: "/admin/quotas/:teamId",
  Settings: "/admin/settings",
  SystemWorkflows: "/admin/system-workflows",
  TaskTemplates: "/admin/task-templates",
  TaskTemplateEdit: `/admin/task-templates/:id/:version`,
  TaskTemplateYaml: `/admin/task-templates/:id/:version/yaml-editor`,
  Team: "/admin/teams/:teamId",
  TeamSettings: "/admin/teams/:teamId/settings",
  TeamWorkflows: "/admin/teams/:teamId/workflows",
  TeamList: "/admin/teams",
  TeamProperties: `/manage/team-parameters`,
  User: "/admin/users/:userId",
  UserList: "/admin/users",

  Workflows: "/workflows",
  WorkflowsMine: "/workflows/mine",
  WorkflowsTeams: "/workflows/teams",

  ManageTaskTemplates: `/manage/task-templates`,
  ManageTaskTemplatesTeam: `/manage/task-templates/team/:teamId`,
  ManageTaskTemplateEdit: `/manage/task-templates/team/:teamId/:taskId/:version`,
  ManageTaskTemplateYaml: `/manage/task-templates/team/:teamId/:taskId/:version/yaml-editor`,
};

export const appLink = {
  activity: () => `/activity`,
  editorDesigner: ({ workflowId }) => `/editor/${workflowId}/workflow`,
  editorConfigure: ({ workflowId }) => `/editor/${workflowId}/configure`,
  editorChangelog: ({ workflowId }) => `/editor/${workflowId}/changelog`,
  editorProperties: ({ workflowId }) => `/editor/${workflowId}/parameters`,
  execution: ({ executionId, workflowId }) => `/activity/${workflowId}/execution/${executionId}`,
  insights: () => "/insights",
  properties: () => "/admin/parameters",
  manageUsers: () => "/admin/users",
  quotas: () => "/admin/quotas",
  quotasEdit: ({ teamId }) => `/admin/quotas/${teamId}`,
  settings: () => "/admin/settings",
  systemWorkflows: () => "/admin/system-workflows",
  taskTemplates: () => "/admin/task-templates",
  taskTemplateEdit: ({ id, version }) => `/admin/task-templates/${id}/${version}`,
  taskTemplateYaml: ({ id, version }) => `/admin/task-templates/${id}/${version}/yaml-editor`,
  teamProperties: () => `/manage/team-parameters`,
  teamTaskTemplates: () => `/manage/task-templates`,
  team: ({ teamId }) => `/admin/teams/${teamId}`,
  teamWorkflows: ({ teamId }) => `/admin/teams/${teamId}/workflows`,
  teamSettings: ({ teamId }) => `/admin/teams/${teamId}/settings`,
  teamList: () => "/admin/teams",
  workflows: () => "/workflows",
  workflowsMine: () => "/workflows/mine",
  workflowsTeams: () => "/workflows/teams",
  workflowActivity: ({ workflowId }) => `/activity?page=0&size=10&workflowIds=${workflowId}`,

  manageTaskTemplates: ({ teamId }) => `/manage/task-templates/team/${teamId}`,
  manageTaskTemplateEdit: ({ teamId, taskId, version }) => `/manage/task-templates/team/${teamId}/${taskId}/${version}`,
  manageTaskTemplateYaml: ({ teamId, taskId, version }) =>
    `/manage/task-templates/team/${teamId}/${taskId}/${version}/yaml-editor`,

  //external apps
  docsWorkflowEditor: () => `${BASE_DOCUMENTATION_URL}/how-to-guide/workflow-editor`,
};

export const queryStringOptions = { arrayFormat: "comma", skipEmptyString: true };

export const FeatureFlag = {
  /**
   * new Feature Flags
   */
  TeamManagementEnabled: "TeamManagementEnabled",
  WorkflowQuotasEnabled: "WorkflowQuotasEnabled",
  SettingsEnabled: "SettingsEnabled",
  UserManagementEnabled: "UserManagementEnabled",
  GlobalParametersEnabled: "GlobalParametersEnabled",
  WorkflowTokensEnabled: "WorkflowTokensEnabled",
  TaskManagerEnabled: "TaskManagerEnabled",
  EditVerifiedTasksEnabled: "EditVerifiedTasksEnabled",
  WorkflowTriggersEnabled: "WorkflowTriggersEnabled",
  TeamParametersEnabled: "TeamParametersEnabled",

  ActivityEnabled: "ActivityEnabled",
  InsightsEnabled: "InsightsEnabled",
};
