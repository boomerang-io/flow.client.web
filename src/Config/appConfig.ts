// Look for the data injected into the HTML file from the Express app
// See server/app.js for implementation
import { Envs } from "Constants";
import { StringifyOptions } from "query-string";
export const APP_ROOT =
  window._SERVER_DATA && window._SERVER_DATA.APP_ROOT ? window._SERVER_DATA.APP_ROOT : "/BMRG_APP_ROOT_CONTEXT";

export const CORE_ENV_URL =
  window._SERVER_DATA && window._SERVER_DATA.CORE_ENV_URL ? window._SERVER_DATA.CORE_ENV_URL : "";

export const BASE_DOCUMENTATION_URL = "https://www.useboomerang.io/docs/boomerang-flow";

//@ts-ignore
export const isDevEnv = import.meta.env.MODE === Envs.Dev;
//@ts-ignore
export const isTestEnv = import.meta.env.MODE === Envs.Test;

type AppPathKey =
  | "Root"
  | "Error"
  | "Activity"
  | "Actions"
  | "ActionsApprovals"
  | "ActionsManual"
  | "Editor"
  | "EditorDesigner"
  | "EditorConfigure"
  | "EditorChangelog"
  | "EditorProperties"
  | "EditorSchedule"
  | "Execution"
  | "Home"
  | "Profile"
  | "Insights"
  | "ManageTaskTemplates"
  | "ManageTaskTemplateDetail"
  | "ManageTaskTemplateEditor"
  | "ManageTeam"
  | "Properties"
  | "Schedules"
  | "Settings"
  | "TemplateWorkflows"
  | "TaskTemplates"
  | "TaskTemplateDetail"
  | "TaskTemplateEditor"
  | "ManageTeam"
  | "ManageTeamSettings"
  | "ManageTeamWorkflows"
  | "ManageTeamLabels"
  | "ManageTeamQuotas"
  | "ManageTeamApprovers"
  | "ManageTeamProperties"
  | "ManageTeamTokens"
  | "Tokens"
  | "TeamList"
  | "User"
  | "UserList"
  | "UserLabels"
  | "UserSettings"
  | "Workflows";

export const AppPath: Record<AppPathKey, string> = {
  Root: "/",
  Error: "/error",
  Activity: "/:teamId/activity",
  Actions: "/:teamId/actions",
  ActionsApprovals: "/:teamId/actions/approvals",
  ActionsManual: "/:teamId/actions/manual",
  Editor: "/:teamId/editor/:workflowId",
  EditorDesigner: `/:teamId/editor/:workflowId/workflow`,
  EditorConfigure: `/:teamId/editor/:workflowId/configure`,
  EditorChangelog: `/:teamId/editor/:workflowId/changelog`,
  EditorProperties: `/:teamId/editor/:workflowId/parameters`,
  EditorSchedule: `/:teamId/editor/:workflowId/schedule`,
  Execution: "/:teamId/activity/:workflowId/execution/:executionId",
  Home: "/home",
  Profile: "/profile",
  Insights: "/:teamId/insights",
  Workflows: "/:teamId/workflows",
  Schedules: "/:teamId/schedules",

  //Manage
  ManageTaskTemplates: `/:teamId/task-templates`,
  ManageTaskTemplateDetail: `/:teamId/task-templates/:name/:version`,
  ManageTaskTemplateEditor: `/:teamId/task-templates/:name/:version/editor`,
  ManageTeamTokens: "/:teamId/tokens",
  ManageTeamProperties: `/:teamId/parameters`,
  ManageTeam: `/:teamId/manage`,
  ManageTeamSettings: "/:teamId/manage/settings",
  ManageTeamWorkflows: "/:teamId/manage/workflows",
  ManageTeamQuotas: "/:teamId/manage/quotas",
  ManageTeamLabels: "/:teamId/manage/Labels",
  ManageTeamApprovers: `/:teamId/manage/approver-groups`,

  //Admin
  Properties: "/admin/parameters",
  Settings: "/admin/settings",
  TemplateWorkflows: "/admin/template-workflows",
  TaskTemplates: "/admin/task-templates",
  TaskTemplateDetail: `/admin/task-templates/:name/:version`,
  TaskTemplateEditor: `/admin/task-templates/:name/:version/editor`,
  TeamList: "/admin/teams",
  Tokens: "/admin/tokens",
  User: "/admin/users/:userId",
  UserLabels: "/admin/users/:userId/labels",
  UserSettings: "/admin/users/:userId/settings",
  UserList: "/admin/users",
};

interface WorkflowIdArg {
  workflowId: string;
}
interface TeamIdArg {
  teamId: string;
}

interface UserIdArg {
  userId: string
}

type TeamRouteArgs = WorkflowIdArg & TeamIdArg;
interface ManageTaskTemplateArgs {
  teamId: string;
  name: string;
  version: string;
}
interface AdminTaskTemplateArgs {
  name: string;
  version: string;
}
interface ExecutionArgs {
  executionId: string;
  workflowId: string;
}

export const appLink = {
  activity: ({ teamId }: TeamIdArg) => `/${teamId}/activity`,
  actions: ({ teamId }: TeamIdArg) => `/${teamId}/actions`,
  actionsApprovals: ({ teamId }: TeamIdArg) => `/${teamId}/actions/approvals`,
  actionsManual: ({ teamId }: TeamIdArg) => `/${teamId}/actions/manual`,
  editorDesigner: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/editor/${workflowId}/workflow`,
  editorConfigure: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/editor/${workflowId}/configure`,
  editorChangelog: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/editor/${workflowId}/changelog`,
  editorProperties: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/editor/${workflowId}/parameters`,
  editorSchedule: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/editor/${workflowId}/schedule`,
  execution: ({ executionId, workflowId }: ExecutionArgs) => `/activity/${workflowId}/execution/${executionId}`,
  home: () => "/home",
  profile: () => "/profile",
  insights: () => "/insights",
  manageTaskTemplates: ({ teamId }: TeamIdArg) =>
    `/${teamId}/task-templates`,
  manageTaskTemplateEdit: ({ teamId, name, version }: ManageTaskTemplateArgs) =>
    `/${teamId}/task-templates/${name}/${version}`,
  manageTaskTemplateYaml: ({ teamId, name, version }: ManageTaskTemplateArgs) =>
    `/${teamId}/task-templates/${name}/${version}/editor`,
  manageTeam: ({ teamId }: TeamIdArg) => `/${teamId}/manage`,
  manageTeamApprovers: ({ teamId }: TeamIdArg) => `/${teamId}/manage/approver-groups`,
  manageTeamWorkflows: ({ teamId }: TeamIdArg) => `/${teamId}/manage/workflows`,
  manageTeamLabels: ({ teamId }: TeamIdArg) => `/${teamId}/manage/labels`,
  manageTeamQuotas: ({ teamId }: TeamIdArg) => `/${teamId}/manage/quotas`,
  manageTeamSettings: ({ teamId }: TeamIdArg) => `/${teamId}/manage/settings`,
  manageTeamTokens: ({ teamId }: TeamIdArg) => `/${teamId}/tokens`,
  manageTeamParameters: ({ teamId }: TeamIdArg) => `/${teamId}/parameters`,
  manageUsers: () => "/admin/users",
  properties: () => "/admin/parameters",
  schedule: () => "/schedule",
  settings: () => "/admin/settings",
  templateWorkflows: () => "/admin/template-workflows",
  taskTemplates: () => "/admin/task-templates",
  taskTemplateDetail: ({ name, version }: AdminTaskTemplateArgs) => `/admin/task-templates/${name}/${version}`,
  taskTemplateEditor: ({ name, version }: AdminTaskTemplateArgs) => `/admin/task-templates/${name}/${version}/editor`,
  teamList: () => "/admin/teams",
  tokens: () => `/admin/tokens`,
  user: ({ userId }: UserIdArg) => `/admin/users/${userId}`,
  userLabels: ({ userId }: UserIdArg) => `/admin/users/${userId}/labels`,
  userSettings: ({ userId }: UserIdArg) => `/admin/users/${userId}/settings`,
  userList: () => "/admin/users",
  workflows: ({ teamId }: TeamIdArg) => `/${teamId}/workflows`,
  workflowActivity: ({ teamId, workflowId }: TeamRouteArgs) => `/${teamId}/activity?page=0&size=10&workflows=${workflowId}`,
  //external apps
  docsWorkflowEditor: () => `${BASE_DOCUMENTATION_URL}/how-to-guide/workflow-editor`,
};

/**
 * new Feature Flags
 */
export enum FeatureFlag {
  ActivityEnabled = "ActivityEnabled",
  EditVerifiedTasksEnabled = "EditVerifiedTasksEnabled",
  GlobalParametersEnabled = "GlobalParametersEnabled",
  InsightsEnabled = "InsightsEnabled",
  TeamManagementEnabled = "TeamManagementEnabled",
  TeamParametersEnabled = "TeamParametersEnabled",
  TeamTasksEnabled = "TeamTasksEnabled",
  UserManagementEnabled = "UserManagementEnabled",
  WorkflowQuotasEnabled = "WorkflowQuotasEnabled",
  WorkflowTokensEnabled = "WorkflowTokensEnabled",
  WorkflowTriggersEnabled = "WorkflowTriggersEnabled",
}

export const queryStringOptions: StringifyOptions = { arrayFormat: "comma", skipEmptyString: true };
