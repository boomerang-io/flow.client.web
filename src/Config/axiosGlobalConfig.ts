// Setup global config for Axios requests so they work with ICP
import axios from "axios";
axios.defaults.withCredentials = true;
