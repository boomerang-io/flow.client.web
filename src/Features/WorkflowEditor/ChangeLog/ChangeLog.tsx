import React from "react";
import { Helmet } from "react-helmet";
import ChangeLogTable from "./ChangeLogTable";
import { ChangeLog as ChangeLogType } from "Types";
import styles from "./changeLog.module.scss";

interface ChangeLogProps {
  changeLogData: ChangeLogType;
}

function ChangeLog({ changeLogData }: ChangeLogProps) {
  return (
    <div className={styles.container}>
      <Helmet>
        <title>Change Log</title>
      </Helmet>
      <ChangeLogTable changeLog={changeLogData} />
    </div>
  );
}

export default ChangeLog;
