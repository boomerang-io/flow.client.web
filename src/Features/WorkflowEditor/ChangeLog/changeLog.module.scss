.container {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  overflow: auto;
  padding: 2rem 2rem 1rem;
  width: 100%;

  :global .bmrg--b-loading-animation__text {
    display: none;
  }
  :global .bmrg--b-loading-animation__img {
    transform: scale(0.3);
  }
}

.searchSkeleton {
  margin-bottom: 1rem;
  width: 300px;
}
