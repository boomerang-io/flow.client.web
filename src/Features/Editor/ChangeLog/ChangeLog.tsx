import React from "react";
import { Helmet } from "react-helmet";
import { useQuery } from "Hooks";
import { DelayedRender, DataTableSkeleton, SearchSkeleton } from "@boomerang-io/carbon-addons-boomerang-react";
import ErrorDragon from "Components/ErrorDragon";
import ChangeLogTable from "./ChangeLogTable";
import qs from "query-string";
import { serviceUrl } from "Config/servicesConfig";
import { WorkflowSummary } from "Types";
import styles from "./changeLog.module.scss";

interface ChangeLogProps {
  summaryData: WorkflowSummary;
}

const ChangeLog: React.FC<ChangeLogProps> = ({ summaryData }) => {
  const getWorkflowChangelogUrl = serviceUrl.getWorkflowChangelog({
    workflowId: summaryData.id,
    query: qs.stringify({ sort: "version", order: "DESC" }),
  });
  const { data, error, isLoading } = useQuery(getWorkflowChangelogUrl);
  if (isLoading)
    return (
      <DelayedRender>
        <div className={styles.container}>
          <div className={styles.searchSkeleton}>
            <SearchSkeleton small />
          </div>
          <DataTableSkeleton />
        </div>
      </DelayedRender>
    );

  if (error) return <ErrorDragon />;

  return (
    <div className={styles.container}>
      <Helmet>
        <title>{`Change Log - ${summaryData.name}`}</title>
      </Helmet>
      <ChangeLogTable changeLog={data} />
    </div>
  );
};

export default ChangeLog;
