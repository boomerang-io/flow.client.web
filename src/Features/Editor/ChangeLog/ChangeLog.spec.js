import React from "react";
import ChangeLog from ".";
import { Route } from "react-router-dom";
import { waitFor } from "@testing-library/react";
import { AppPath, appLink } from "Config/appConfig";
import { startApiServer } from "ApiServer";

let server;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
});

const props = {
  summaryData: {
    id: "5eb2c4085a92d80001a16d87",
    name: "Test name",
  },
};

describe("ChangeLog --- Snapshot Test", () => {
  it("Capturing Snapshot of ChangeLog", async () => {
    const { baseElement, findByText } = rtlContextRouterRender(
      <Route path={AppPath.EditorChangelog}>
        <ChangeLog {...props} />
      </Route>,
      { route: appLink.editorChangelog({ workflowId: "5eb2c4085a92d80001a16d87" }) }
    );
    await findByText("hello sir");
    expect(baseElement).toMatchSnapshot();
    await waitFor(() => {});
  });
});
