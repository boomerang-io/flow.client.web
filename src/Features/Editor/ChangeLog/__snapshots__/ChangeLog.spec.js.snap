// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeLog --- Snapshot Test Capturing Snapshot of ChangeLog 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <div
        class="tableContainer"
      >
        <div
          aria-labelledby="change-log-table-search-search"
          class="bx--search bx--search--xl search"
          role="search"
        >
          <svg
            aria-hidden="true"
            class="bx--search-magnifier"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5	C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
            />
          </svg>
          <label
            class="bx--label"
            for="change-log-table-search"
            id="change-log-table-search-search"
          >
            Search
          </label>
          <input
            autocomplete="off"
            class="bx--search-input"
            data-testid="change-log-search"
            id="change-log-table-search"
            placeholder="Search"
            role="searchbox"
            type="text"
          />
          <button
            aria-label="Clear search input"
            class="bx--search-close bx--search-close--hidden"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
              />
            </svg>
          </button>
        </div>
        <div
          class="bx--data-table-container"
        >
          <div
            class="bx--data-table-content"
          >
            <table
              class="bx--data-table bx--data-table--sort bx--data-table--no-border"
            >
              <thead>
                <tr>
                  <th
                    aria-sort="none"
                    scope="col"
                  >
                    <button
                      class="bx--table-sort"
                      id="version"
                      type="button"
                    >
                      <span
                        class="bx--table-sort__flex"
                      >
                        <div
                          class="bx--table-header-label"
                        >
                          Version
                        </div>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    scope="col"
                  >
                    <button
                      class="bx--table-sort"
                      id="userName"
                      type="button"
                    >
                      <span
                        class="bx--table-sort__flex"
                      >
                        <div
                          class="bx--table-header-label"
                        >
                          User
                        </div>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    scope="col"
                  >
                    <button
                      class="bx--table-sort"
                      id="reason"
                      type="button"
                    >
                      <span
                        class="bx--table-sort__flex"
                      >
                        <div
                          class="bx--table-header-label"
                        >
                          Reason
                        </div>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                  <th
                    aria-sort="none"
                    scope="col"
                  >
                    <button
                      class="bx--table-sort"
                      id="date"
                      type="button"
                    >
                      <span
                        class="bx--table-sort__flex"
                      >
                        <div
                          class="bx--table-header-label"
                        >
                          Date
                        </div>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                          />
                        </svg>
                        <svg
                          aria-label="Sort rows by this header in ascending order"
                          class="bx--table-sort__icon-unsorted"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                          />
                        </svg>
                      </span>
                    </button>
                  </th>
                </tr>
              </thead>
              <tbody
                aria-live="polite"
                class="tableBody"
              >
                <tr
                  data-testid="change-log-table-row"
                >
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        2
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        Boomerang/chicago/ibm Boomerang/chicago/ibm
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        Sorry. Removing changes.
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        2020-01-01 12:00 AM
                      </p>
                    </div>
                  </td>
                </tr>
                <tr
                  data-testid="change-log-table-row"
                >
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        1
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        Boomerang/chicago/ibm Boomerang/chicago/ibm
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        hello sir
                      </p>
                    </div>
                  </td>
                  <td>
                    <div
                      class="tableCell"
                    >
                      <p
                        class="tableText"
                      >
                        2020-01-01 12:00 AM
                      </p>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="bx--pagination"
        >
          <div
            class="bx--pagination__left"
          >
            <label
              class="bx--pagination__text"
              for="bx-pagination-select-1"
              id="bx-pagination-select-1-count-label"
            >
              Items per page:
            </label>
            <div
              class="bx--form-item"
            >
              <div
                class="bx--select bx--select--inline bx--select__item-count"
              >
                <div
                  class="bx--select-input--inline__wrapper"
                >
                  <div
                    class="bx--select-input__wrapper"
                  >
                    <select
                      class="bx--select-input"
                      id="bx-pagination-select-1"
                    >
                      <option
                        class="bx--select-option"
                        value="10"
                      >
                        10
                      </option>
                      <option
                        class="bx--select-option"
                        value="25"
                      >
                        25
                      </option>
                      <option
                        class="bx--select-option"
                        value="50"
                      >
                        50
                      </option>
                    </select>
                    <svg
                      aria-hidden="true"
                      class="bx--select__arrow"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="bx--pagination__text bx--pagination__items-count"
            >
              1–2 of 2 items
            </span>
          </div>
          <div
            class="bx--pagination__right"
          >
            <div
              class="bx--form-item"
            >
              <div
                class="bx--select bx--select--inline bx--select__page-number"
              >
                <label
                  class="bx--label bx--visually-hidden"
                  for="bx-pagination-select-1-right"
                >
                  Page number, of 1 pages
                </label>
                <div
                  class="bx--select-input--inline__wrapper"
                >
                  <div
                    class="bx--select-input__wrapper"
                  >
                    <select
                      class="bx--select-input"
                      id="bx-pagination-select-1-right"
                    >
                      <option
                        class="bx--select-option"
                        value="1"
                      >
                        1
                      </option>
                    </select>
                    <svg
                      aria-hidden="true"
                      class="bx--select__arrow"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      preserveAspectRatio="xMidYMid meet"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <span
              class="bx--pagination__text"
            >
              of 1 page
            </span>
            <div
              class="bx--pagination__control-buttons"
            >
              <button
                class="bx--pagination__button bx--pagination__button--backward bx--pagination__button--no-index bx--btn bx--btn--ghost bx--btn--disabled bx--btn--icon-only bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--top bx--tooltip--align-center"
                disabled=""
                tabindex="0"
                type="button"
              >
                <span
                  class="bx--assistive-text"
                >
                  Previous page
                </span>
                <svg
                  aria-hidden="true"
                  aria-label="Previous page"
                  class="bx--btn__icon"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20 24L10 16 20 8z"
                  />
                </svg>
              </button>
              <button
                class="bx--pagination__button bx--pagination__button--forward bx--pagination__button--no-index bx--btn bx--btn--ghost bx--btn--disabled bx--btn--icon-only bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--top bx--tooltip--align-end"
                disabled=""
                tabindex="0"
                type="button"
              >
                <span
                  class="bx--assistive-text"
                >
                  Next page
                </span>
                <svg
                  aria-hidden="true"
                  aria-label="Next page"
                  class="bx--btn__icon"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 8L22 16 12 24z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
